File: Programação Dinâmica/G.cpp
#include <bits/stdc++.h>

using namespace std;

using llu = long long unsigned int;

int main(){
    int n, w;
    cin >> n >> w;

    vector<int> pesos(n+1);
    vector<int> valor(n+1);
    for(int i = 1; i <= n; i++){
        cin >> pesos[i] >> valor[i];
    }
    vector<vector<llu>> dp(n+1, vector<llu>(w+1, 0));

    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= w; j++){
            if(pesos[i] <= j)
                dp[i][j] = max(dp[i-1][j], dp[i-1][j-pesos[i]] + valor[i]);
            else
                dp[i][j] = dp[i-1][j];
        }
    }

    cout << dp[n][w] << "\n";
}

File: Programação Dinâmica/G.py
n, w = map(int, input().split())

pesos = [0]
valor = [0]
for _ in range(n):
    p, v = map(int, input().split())
    pesos.append(p)
    valor.append(v)

dp = [[0 for _ in range(w+1)] for _ in range(n+1)]

for i in range(1, n+1):
    for j in range(1, w+1):
        if pesos[i] <= j:
            dp[i][j] = max(dp[i-1][j], dp[i-1][j-pesos[i]] + valor[i])
        else:
            dp[i][j] = dp[i-1][j]
        
print(dp[n][w])